magento/magento2
# Improve catalog rule product indexing performance

### Description (*)

Updates the IndexBuilder class to call `fetchItem` on the Collection rather than `foreach`. This prevents the entire collection being loaded at once which significantly speeds up the indexing time

### Fixed Issues (if relevant)

1. magento/magento2#25889: Catalog Rule Product index time increases exponentially as new rules are added

### Manual testing scenarios (*)

1. Install Magento
2. Copy setup/performance-toolkit/profiles/ce/small.xml to a new file
3. Update the file to have 2,000 simple products, 2,000 categories, and 25,000 catalog price rules
4. Generate fixtures using the file
5. Run the Catalog Rule Product index
6. The index should complete in a few minutes rather than an hour

### Questions or comments

Alongside the change mentioned in the ticket, there are a couple of other places in the class that were calling `foreach` on the collection. I've applied the same update to them as well
 
I'm not sure of the best way to put together a test to cover these changes, the behaviour should be the same just faster. If you can give any guidance on what should be included I'll add in tests for this.

### Contribution checklist (*)
